File: Java/leetcode75/src/main/java/org/phoenix/leetcode/challenges/Problem34_LeafSimilarTrees.java
package org.phoenix.leetcode.challenges;

import org.phoenix.leetcode.challenges.Problem33_MaximumDepthOfBinaryTree.TreeNode;

import java.util.ArrayList;
import java.util.List;

public class Problem34_LeafSimilarTrees {
    public boolean leafSimilar(TreeNode root1, TreeNode root2) {
        List<Integer> left = new ArrayList<>();
        List<Integer> right = new ArrayList<>();
        dfs(root1, left);
        dfs(root2, right);
        return left.equals(right);
    }

    public void dfs(TreeNode root, List<Integer> leafList) {
        if (root == null) return;
        if (root.left == null && root.right == null) {
            leafList.add(root.val);
        }
        dfs(root.left, leafList);
        dfs(root.right, leafList);
    }
}

File: JavaScript/leetcode75/src/main/Problem34_LeafSimilarTrees.js
/**
 * @param {TreeNode} root1
 * @param {TreeNode} root2
 * @return {boolean}
 */
var leafSimilar = function (root1, root2) {
  let arr1 = [];
  let arr2 = [];
  dfs(root1, arr1);
  dfs(root2, arr2);
  return arr1.toString() === arr2.toString();
};

var dfs = function (root, arr) {
  if (root === null) return;
  if (!root.left && !root.right) {
    arr.push(root.val);
  }
  dfs(root.left, arr);
  dfs(root.right, arr);
};

module.exports = leafSimilar;
